File: api-gateway-java/src/main/java/ru/kuramshindev/apigatewayjava/GatewayServiceImpl.java
package ru.kuramshindev.apigatewayjava;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import ru.kuramshin.dto.apigateway.ApiGatewayDto;
import ru.kuramshin.dto.companies.CompanyDto;
import ru.kuramshin.dto.contractors.ContractorDto;
import ru.kuramshin.dto.projects.ProjectDto;

import java.util.UUID;
import java.util.concurrent.ExecutionException;

@Slf4j
@RequiredArgsConstructor
@Service("gatewayServiceImpl")
public class GatewayServiceImpl implements GatewayService {

    private final UpstreamClients upstreamClients;

    @Override
    public ApiGatewayDto getContractById(UUID uuid) {

        // Получение компании
        CompanyDto company = upstreamClients.getCompanyById(uuid);

        // Получение подрядчика
        ContractorDto contractor = upstreamClients.getContractorById(uuid);

        // Получение проекта
        ProjectDto project = upstreamClients.getProjectById(uuid);

        // Собираем ответ
        return ApiGatewayDto.builder()
                .company(company)
                .contractor(contractor)
                .project(project)
                .build();
    }
}

File: api-gateway-kotlin-reactive/src/main/kotlin/ru/kuramshindev/apigatewaykotlinreactive/GatewayServiceImpl.kt
package ru.kuramshindev.apigatewaykotlinreactive

import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import org.springframework.stereotype.Service
import ru.kuramshin.dto.apigateway.ApiGatewayDto
import java.util.*

@Service
class GatewayServiceImpl(
    private val upstreamAsyncClients: UpstreamAsyncClients
) : GatewayService {

    override suspend fun getContractById(uuid: UUID): ApiGatewayDto = coroutineScope {

        // Получение компании
        val company = async { upstreamAsyncClients.getCompanyByIdAsync(uuid) }

        // Получение подрядчика
        val contractor = async { upstreamAsyncClients.getContractorByIdAsync(uuid) }

        // Получение проекта
        val project = async { upstreamAsyncClients.getProjectByIdAsync(uuid) }

        // Собираем ответ
        ApiGatewayDto.builder()
            .company(company.await())
            .contractor(contractor.await())
            .project(project.await())
            .build()
    }
}